### PR TITLE
Js_of_ocaml TodoMVC Example.

### PR DESCRIPTION
**Pull request by slegrand45**
[Js_of_ocaml](http://ocsigen.org/js_of_ocaml/) is a compiler of OCaml bytecode to Javascript.

[OCaml](https://ocaml.org/) is a general purpose programming language with an emphasis on expressiveness and safety. Js_of_ocaml brings nearly all the OCaml features to Javascript. And especially its powerful type system.

### CODE REVIEW

**Comment by arthurvr** [comment]
Personally I think should totally go discuss this at [TasteLang](https://github.com/tastejs/TasteLang) but that it isn't really a good fit for TodoMVC. Mainly because this is more about the language than about an MV* framework. What does the rest think?

**Comment by passy** [comment]
OCaml has a really nice JS story these days and I hear it more and more, especially among people looking for a type-safe way to write web apps, where there aren't very many options out there. The implementation does a pretty nice job of separating the concerns and looks pretty similar to the Elm way, so I'm :+1:.

File: examples/js_of_ocaml/build.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+ocamlbuild -use-ocamlfind \
+  -pkgs lwt.syntax,js_of_ocaml,js_of_ocaml.syntax,js_of_ocaml.tyxml,tyxml,js_of_ocaml.deriving,js_of_ocaml.deriving.syntax,deriving \
+  -syntax camlp4o \
+  todomvc.byte
+
+js_of_ocaml +weak.js --opt 3 -o js/todomvc.js todomvc.byte
+
+# Avoid JS linters (hard to apply for auto generated JS)
```
Comment (arthurvr) [inline examples/js_of_ocaml/build.sh]: Instead of this magic just ignore specific files in the config files.

File: examples/js_of_ocaml/index.html [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Js_of_ocaml • TodoMVC</title>
+    <link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+  </head>
+  <body>
+    <div id="todomvc"></div>
+    <script src="js/todomvc.js"></script>
+  </body>
+</html>
```
Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: missing newline

File: examples/js_of_ocaml/index.html [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
+  <head>
```
Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: we use tabs for indentation.

File: examples/js_of_ocaml/index.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Js_of_ocaml • TodoMVC</title>
+    <link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
```
Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: This misses todomvc-common (check the other apps)

File: examples/js_of_ocaml/index.html [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<!doctype html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>Js_of_ocaml • TodoMVC</title>
```
Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: Let's properly write this instead of snake_case.

**Comment by arthurvr** [comment]
Any active community member who's able to review the code? I don't think any of us is (or am I wrong?)

File: examples/js_of_ocaml/index.html (hunk 5 above)
Comment (passy) [inline examples/js_of_ocaml/index.html]: I think this is the official spelling.

Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: Oh. Alright. Pretty ugly :smile_cat: Thanks @passy!

**Comment by passy** [comment]
I'm at a read-only level of OCaml understanding. I'll give it a more
thorough look, but it would indeed be great to have someone with actual
experience in this space to review.

**Comment by slegrand45** [comment]
The code has been reviewed by @Drup who is a member of the jsoo project. The code style is now much better.

**Comment by passy** [comment]
Oh yes, I prefer the new style a lot over the old `let () ... in`. Well done! :)

File: examples/js_of_ocaml/.gitignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+_build
+*.byte
```
Comment (arthurvr) [inline examples/js_of_ocaml/.gitignore]: missing a newline

File: examples/js_of_ocaml/.jscsrc [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	"disallowYodaConditions": true,
+	"disallowNewlineBeforeBlockStatements": true,
+	"maximumLineLength": 120,
+	"validateLineBreaks": "LF",
+	"validateIndentation": "\t",
+	"disallowKeywords": ["with"],
+	"disallowSpacesInsideObjectBrackets": null,
+	"disallowImplicitTypeConversion": ["string"],
+	"safeContextKeyword": "self"
+}
```
Comment (arthurvr) [inline examples/js_of_ocaml/.jscsrc]: Why is this file here? Just use our code style guidelines and just use the `jscsrc` in the root directory.

File: examples/js_of_ocaml/.jshintignore [hunk 8]
```diff
@@ -0,0 +1 @@
+js/todomvc.ml
```
Comment (arthurvr) [inline examples/js_of_ocaml/.jshintignore]: Isn't there a better way to ignore files then per-example `jshintignore` files? @samccone any ideaz?

File: examples/js_of_ocaml/build.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+js_of_ocaml +weak.js --opt 3 -o js/todomvc.js todomvc.byte
+
+
+# To disable JS linters at a file level, you can for instance use sed:
+# sed -i '' '1i\
+# // jscs:disable
+# ' js/todomvc.js
+# sed -i '' '1i\
+# /* jshint ignore:start */
+# ' js/todomvc.js
```
Comment (arthurvr) [inline examples/js_of_ocaml/build.sh]: Commented out code doesn't make sense. Let's remove it together with lines 9 and 10.

File: examples/js_of_ocaml/build.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
+ocamlbuild -use-ocamlfind \
```
Comment (arthurvr) [inline examples/js_of_ocaml/build.sh]: this could use a comment to explain what it does.

File: examples/js_of_ocaml/index.html [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+		<title>Js_of_ocaml • TodoMVC</title>
+		<link rel="stylesheet" href="node_modules/todomvc-common/base.css">
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+	</head>
+	<body>
+		<div id="todomvc"></div>
+		<script src="js/todomvc.js"></script>
+	</body>
+</html>
+
```
Comment (arthurvr) [inline examples/js_of_ocaml/index.html]: remove this empty line.

File: examples/js_of_ocaml/todomvc.ml [hunk 12]
```diff
@@ -0,0 +1,423 @@
+
```
Comment (arthurvr) [inline examples/js_of_ocaml/todomvc.ml]: remove dat line (or is it there for a reason?)

**Comment by arthurvr** [comment]
Neat work. Thanks.

File: examples/js_of_ocaml/.jscsrc (hunk 7 above)
Comment (slegrand45) [inline examples/js_of_ocaml/.jscsrc]: I copied the `jscsrc` from the root directory to be able to modify the `excludeFiles` directive:
```
"excludeFiles": ["**/node_modules/**", "**/bower_components/**", "**/js/todomvc.js"],
```

This is the only method i found to exclude the js file from the `jscsrc` linter without directly putting a directive at the top of the js file. Is there any better way to exclude a file from `jscsrc`? It seems that `jscsrc` doesn't support a feature similar to the `.jshintignore` file.

File: examples/js_of_ocaml/.jshintignore (hunk 8 above)
Comment (samccone) [inline examples/js_of_ocaml/.jshintignore]: hmm wow i have never seen jshintignore but it actually seems perfect for what we want!

Comment (samccone) [inline examples/js_of_ocaml/.jshintignore]: we just need to have our lint task actually read from this file when linting

File: examples/js_of_ocaml/.jscsrc (hunk 7 above)
Comment (arthurvr) [inline examples/js_of_ocaml/.jscsrc]: Just add the ignore in the file in the root?

File: examples/js_of_ocaml/readme.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+2. See the `build.sh` script.
+
+
+## Credit
+
+Created by [Stéphane Legrand](https://stephanelegrand.wordpress.com/).
+
+Various code improvements from [Gabriel Radanne](https://github.com/Drup).
+
+Based on [Elm implementation](https://github.com/evancz) by Evan Czaplicki.
```
Comment (arthurvr) [inline examples/js_of_ocaml/readme.md]: >  https://github.com/evancz

make this a link on the name itself

File: examples/js_of_ocaml/.jscsrc (hunk 7 above)
Comment (slegrand45) [inline examples/js_of_ocaml/.jscsrc]: If it's possible to propose a modification of the `.jscsrc` file (https://github.com/tastejs/todomvc/blob/master/.jscsrc) then maybe it could be something more generic? Something like:

```
"excludeFiles": ["**/node_modules/**", "**/bower_components/**", "**/auto_gen/**"],
```

So if an implementation uses some automatically generated js files, it can put them in this directory to prevent the `jscsrc` checkings.

Comment (arthurvr) [inline examples/js_of_ocaml/.jscsrc]: Nah. Just add the directory you don't want to check in the jscsrc in the root directory and remove this file :)

File: examples/js_of_ocaml/.jshintignore (hunk 8 above)
Comment (arthurvr) [inline examples/js_of_ocaml/.jshintignore]: :+1:

File: examples/js_of_ocaml/build.sh [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+# Compile OCaml source file to OCaml bytecode
+ocamlbuild -use-ocamlfind \
+  -tags "warn(-40)" \
+  -pkgs lwt.syntax,js_of_ocaml,js_of_ocaml.syntax,js_of_ocaml.tyxml,tyxml,js_of_ocaml.deriving,js_of_ocaml.deriving.syntax,deriving \
+  -syntax camlp4o \
+  todomvc.byte ;
+
+# Build JS code from the OCaml bytecode
+js_of_ocaml +weak.js --opt 3 -o js/todomvc.js todomvc.byte
```
Comment (arthurvr) [inline examples/js_of_ocaml/build.sh]: Can you add a description to the readme about how to bulid the app? That makes it clear for people and helps us to maintain the app further on.

File: examples/js_of_ocaml/build.sh [hunk 15]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+# Compile OCaml source file to OCaml bytecode
+ocamlbuild -use-ocamlfind \
+  -tags "warn(-40)" \
```
Comment (arthurvr) [inline examples/js_of_ocaml/build.sh]: Wrong indentation. Check the editorconfig :)

File: examples/js_of_ocaml/readme.md [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+
+## Implementation
+
+Open `index.html` in your browser to try the application.
+
+If you want to build the application on your own:
+
+1. Install Js_of_ocaml. An easy way is to use [opam](https://opam.ocaml.org/).
+2. See the `build.sh` script.
```
Comment (arthurvr) [inline examples/js_of_ocaml/readme.md]: I'd like to extend this description a bit on how exactly to run the `build.sh` script.

File: examples/js_of_ocaml/readme.md [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+
+## Credit
+
+Created by [Stéphane Legrand](https://stephanelegrand.wordpress.com/).
```
Comment (arthurvr) [inline examples/js_of_ocaml/readme.md]: Try to be consistent. Here you're linking to your personal site while for other people you link to their GitHub profile :)

File: examples/js_of_ocaml/readme.md [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+> Js_of_ocaml is a compiler of OCaml bytecode to Javascript.
+
+## Resources
+
+- [Website](http://ocsigen.org/js_of_ocaml/)
+- [GitHub](https://github.com/ocsigen/js_of_ocaml)
+- [Try Js_of_ocaml](http://try.ocamlpro.com/js_of_ocaml/)
+
+### Support
```
Comment (arthurvr) [inline examples/js_of_ocaml/readme.md]: `##`

**Comment by arthurvr** [comment]
Thanks for your nice work!

File: examples/js_of_ocaml/todomvc.ml (hunk 12 above)
Comment (slegrand45) [inline examples/js_of_ocaml/todomvc.ml]: I think it's more readable with this separation line because it clearly separates the `open` directive from the modules. But i guess it's a matter of taste.

File: examples/js_of_ocaml/index.html [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+<html lang="en">
+	<head>
+		<meta charset="utf-8">
+		<meta name="viewport" content="width=device-width, initial-scale=1">
+		<title>Js_of_ocaml • TodoMVC</title>
+		<link rel="stylesheet" href="node_modules/todomvc-common/base.css">
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+	</head>
+	<body>
+		<div id="todomvc"></div>
```
Comment (samccone) [inline examples/js_of_ocaml/index.html]: it looks like you are using the 2.0 spec however this root element is an ID, can you switch it to a class?

Comment (slegrand45) [inline examples/js_of_ocaml/index.html]: Is this an absolute requirement? The standard way with js_of_ocaml is to select first an element by its id and then use it to plug the application. I could select the `<div>` element with the `getElementsByClassName` method but unfortunately this method is for now only available in the development version of js_of_ocaml.

Comment (samccone) [inline examples/js_of_ocaml/index.html]: Is this an absolute requirement? I do not think so, however our test suite does a lookup on this element to determine what version of the spec to use when testing, so clearly this does pose some issues.

What do you think?

Comment (slegrand45) [inline examples/js_of_ocaml/index.html]: Could we have an `id` attribute and a `class` attribute? ie something like:
```html
<div id="todomvc" class="todoapp"></div>
```

Otherwise, if the idea is to ban any `id` attribute, i have no solution for now until the next release of js_of_ocaml (which will include the `getElementsByClassName` method).

By the way, can you confirm that the needed class name is `todoapp`? Thanks.

**Comment by samccone** [comment]
we look for `todoapp` here is the commit that adds the logic.
https://github.com/tastejs/todomvc/commit/45b15b5428bf2bd4313cd7b7aa31ad73c1327be6

For now doing both is an OK solution, thanks for chatting with me, and thanks for the awesome work.

--

ps I tried compiling this app locally by following your instructions and ran into all kinds of errors when installing dependencies, Do you think you might be able to expand the setup instructions for how to compile this app, thanks!

**Comment by slegrand45** [comment]
I added the "todoapp" class. Thank you for your feedback about your attempt to compile the application. I made a test by starting from a clean environment. And indeed it's not as simple as it seems... So i updated the building instructions. I think it's a lot better now.

**Comment by samccone** [comment]
looks like we are in need of a rebase :globe_with_meridians:

**Comment by slegrand45** [comment]
After some fixes, there are still 3 failed checks. They are all from the routing part and they all fail because of "stale element reference: element is not attached to the page document". Not sure if i can apply some workarounds?

**Comment by samccone** [comment]
@slegrand45 how recently have you rebased off of master? We Very recently merged a fix https://github.com/tastejs/todomvc/pull/1371

**Comment by arthurvr** [comment]
Seems like some unrelated commits got into this PR. Can you rebase and get them out?

**Comment by slegrand45** [comment]
It's strange because i already made a rebase... Anyway, i used the following commands:
```sh
git checkout js_of_ocaml
git fetch upstream
git rebase upstream/master
git push -f
```
Strangely enough, i had several conflicts about my own commits during the rebase (?). I used --skip to resolve them. I had also to use "push -f" because git said that my local branch and "origin/js_of_ocaml" have diverged (?). It seems that the PR is now clean again. But i hope i didn't introduce some mess :worried:

Besides, the CI log still shows some "stale element reference" (?).

**Comment by samccone** [comment]
yep @slegrand45 looks like you uncovered some more flake :clap:

**Comment by samccone** [comment]
hey @slegrand45 would you mind rebasing this off of master and squashing your commits down?

**Comment by slegrand45** [comment]
Seems ok!

**Comment by samccone** [comment]
@passy I will let you have final :+1: on this

**Comment by passy** [comment]
@slegrand45 I'm seeing some odd rendering errors among other things. Is the latest version on your branch up to date?

screenshot from 2015-10-09 20-36-54

**Comment by slegrand45** [comment]
@passy Yes it's the latest version. What are the errors you are seeing?

**Comment by passy** [comment]
Rendering is broken like on the screenshot and I've only tried a few things that all weren't quite right, focus lost after creating an item, no selection when entering edit mode etc. Do you see those things on your end too?

**Comment by slegrand45** [comment]
Ok, i see what you mean. Indeed, i have the same problems. I'm working on it right now to fix it.

**Comment by slegrand45** [comment]
@passy I think it should be better now.

**Comment by passy** [comment]
@slegrand45 Oh yes, way better! :cactus: I'm checking it now. :)

**Comment by passy** [comment]
Passing the integration tests with flying colors! Checking with `drool` now.

screenshot from 2015-10-10 20-40-39

runningcolourdog

**Comment by passy** [comment]
I took me way longer than it should have to find this GIF, but it was worth it and I'm super excited about landing this. Thanks so much @slegrand45! Also thanks @arthurvr and @samccone for the review! :heart_eyes_cat:

**Comment by slegrand45** [comment]
Thank you very much to all the team! :smiley:

**Comment by LouisAyroles** [comment]
Hello everyone, note sure if it's the right place to ask this :/

Do you have an example with version of dependencies ?
Because I have serious problem to make it work..

When I'm trying with latest opam,eliom config I have problem with js_of_ocaml.deriving.syntax not found.

I've tried many different ways and I'm still stucked..

Thx for yout help. :smile:

**Comment by slegrand45** [comment]
Hi,

Indeed, some packages have been renamed since 2015. I will try to send a PR to fix that. Meanwhile, you can take a look at the examples at [https://github.com/slegrand45/examples_ocsigen](https://github.com/slegrand45/examples_ocsigen). These examples should compile and run without any problem. If not, please send an issue report.
